### PR TITLE
ekf2_main: update stored mag bias variances

### PR DESCRIPTION
**Pull request by bresch**
Instead of using a fixed mag bias variance to compute the influence of the currently estimated mag bias on the previously saved parameter, we now also save the variance of the parameter.

The value is automatically updated and the only remaining tuning parameter is `EKF2_MAGB_K`, which is more intuitive.

### CODE REVIEW

File: src/modules/ekf2/EKF2.cpp [hunk 1]
```diff
@@ -225,20 +225,23 @@ int EKF2::print_status()
 	return 0;
 }
 
-template<typename Param>
-void EKF2::update_mag_bias(Param &mag_bias_param, int axis_index)
+template<typename Param1, typename Param2>
+void EKF2::update_mag_bias(Param1 &mag_bias_param, Param2 &mag_bias_var_param, int axis_index)
```
Comment (dagar) [inline src/modules/ekf2/EKF2.cpp:229]: FYI this is a relatively large amount of template code bloat for fairly simple functionality.

We should refactor this later to make the logic a simple method and handle the param commit separately.

File: src/modules/ekf2/ekf2_params.c [hunk 2]
```diff
@@ -1117,15 +1117,34 @@ PARAM_DEFINE_FLOAT(EKF2_MAGBIAS_Z, 0.0f);
 PARAM_DEFINE_INT32(EKF2_MAGBIAS_ID, 0);
 
 /**
- * State variance assumed for magnetometer bias storage.
- * This is a reference variance used to calculate the fraction of learned magnetometer bias that will be used to update the stored value. Smaller values will make the stored bias data adjust more slowly from flight to flight. Larger values will make it adjust faster.
+ * State variance for magnetometer x-axis bias storage
+ * This is a reference variance used to calculate the fraction of learned magnetometer bias that will be used to update the stored value.
+ *
+ * @group EKF2
+ * @unit mgauss^2
```
Comment (dagar) [inline src/modules/ekf2/ekf2_params.c:1124]:
```suggestion
 * @unit mgauss^2
 * @volatile
```

Any parameters that are updated by the system in normal usage like this should be marked to volatile. This excludes them from the parameter hash used by QGC and can prevent having to resync the entire parameter set every new session.

**Comment by dagar** [comment]
This looks good to me, although it does complicate how I was considering scaling this to a multi-ekf world.

**Comment by dagar** [comment]
Side question, realistically is anyone actually tuning `EKF2_MAGB_K`?

**Comment by bresch** [comment]
> Side question, realistically is anyone actually tuning EKF2_MAGB_K?

Probably only a few people know that this parameter exists. I don't think anyone is actually tuning it.

**Comment by dagar** [comment]
> > Side question, realistically is anyone actually tuning EKF2_MAGB_K?
> 
> Probably only a few people know that this parameter exists. I don't think anyone is actually tuning it.

I suppose it would be silly to suggest it now, but as a more general statement overall I think we'd be better off only introducing parameters that are absolutely needed and going to be used. Otherwise it's yet another dimension of a test matrix we're neglecting, not to mention a small waste of flash, memory, and parameter sync time.

File: src/modules/ekf2/EKF2.cpp [hunk 3]
```diff
@@ -416,13 +419,20 @@ void EKF2::Run()
...
-					// this means we need to reset the learned bias values to zero
+					// this means we need to reset the learned bias values to zero and
+					// the learned bias variance to a default value
 					_param_ekf2_magbias_x.set(0.f);
 					_param_ekf2_magbias_x.commit_no_notification();
 					_param_ekf2_magbias_y.set(0.f);
 					_param_ekf2_magbias_y.commit_no_notification();
 					_param_ekf2_magbias_z.set(0.f);
 					_param_ekf2_magbias_z.commit_no_notification();
+					_param_ekf2_magb_var_x.set(2.5e-7f);
```
Comment (dagar) [inline src/modules/ekf2/EKF2.cpp:430]: We could add a reset (to default) instead. https://github.com/PX4/Firmware/blob/master/platforms/common/include/px4_platform_common/param.h

File: src/modules/ekf2/EKF2.cpp [hunk 4]
```diff
@@ -225,20 +225,23 @@ int EKF2::print_status()
...
 	if (_valid_cal_available[axis_index]) {
 
 		// calculate weighting using ratio of variances and update stored bias values
-		const float weighting = constrain(_param_ekf2_magb_vref.get() / (_param_ekf2_magb_vref.get() +
+		const float weighting = constrain(mag_bias_var_param.get() / (mag_bias_var_param.get() +
 						  _last_valid_variance[axis_index]), 0.0f, _param_ekf2_magb_k.get());
 		const float mag_bias_saved = mag_bias_param.get();
 
 		_last_valid_mag_cal[axis_index] = weighting * _last_valid_mag_cal[axis_index] + mag_bias_saved;
+		const float new_ref_var = weighting * _last_valid_variance[axis_index] + (1.f - weighting) * mag_bias_var_param.get();
```
Comment (MaEtUgR) [inline src/modules/ekf2/EKF2.cpp:239]: Reuse `AlphaFilter<float>[]` for `_last_valid_variance`  https://github.com/PX4/ecl/blob/99aafbaa18986342baf2f69c4fc507a76a8d76b2/AlphaFilter/AlphaFilter.hpp#L93 ?

Comment (bresch) [reply src/modules/ekf2/EKF2.cpp:239]: Well, it isn't worth instantiating a filter just for a single update per flight

Comment (MaEtUgR) [reply src/modules/ekf2/EKF2.cpp:239]: Ah I see 👍

**Comment by MaEtUgR** [comment]
Should we move on with this? What's missing? Templating will be solved in the next step and the parameters added are necessary to keep state. `EKF2_MAGB_K` was not touched and could still get removed.

**Comment by bresch** [comment]
> Should we move on with this? What's missing? Templating will be solved in the next step and the parameters added are necessary to keep state. EKF2_MAGB_K was not touched and could still get removed.

Maybe just the "reset to default": https://github.com/PX4/Firmware/pull/15772#discussion_r492156574 but I'm not sure how to implement it properly, @dagar could you help me with that?
For the rest @priseborough agreed with the idea and suggested that we keep that calibration separated from the `CAL_MAG_...` to avoid modifying the "factory calibration". Also, moving the logic to sensors will require to pass quite a lot of information back to the sensor module, so I think we need to keep it at a higher level since it needs some context. However, maybe a "higher level" isn't "here" given that we're going to switch to a multi-ekf structure soon.

File: src/modules/ekf2/EKF2.cpp (hunk 3 above)
Comment (bresch) [reply src/modules/ekf2/EKF2.cpp:430]: That would be better, yes. Could you help me to do that "reset to default" function please?

**Comment by bresch** [comment]
After more reflection, I think this change won't improve the estimate that much as we would need to add some process noise to factor in the bias deviation over time to be more accurate. We would then need to know the current age of the drone and since the changes in mag bias are more often due to magnetization, they are not slowly changing over time and adding some process noise won't be much better than the current implementation with a fix variance.

**Pull request closed without merging**